git-user-count/user-count.sh

```diff
@@ -0,0 +1,138 @@
+#!/bin/bash
+set -e
+
+basedir="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
+repositories=$basedir/repositories.txt
+userListFile="$basedir/gradle-enterprise-users.txt"
+userUniqueListFile="$basedir/gradle-enterprise-unique-users.txt"
+usersByRepoFile="$basedir/gradle-enterprise-users-by-repo.csv"
+checkout_area=.tmp/repos
+since=30
+shallow_clone=true
+git_options=
+
+yellow='\033[1;33m'
+nc='\033[0m'
+
+# -b option lets you specify a branch to checkout, store it in a variable
+# -s option lets you specify a number of days to count commits from, store it in a variable otherwise default to 30
+# -c option disables shallow cloning
+# -o option lets you specify aditional git cloning options
+while getopts b:s:co: option
+do
+  case "${option}" in
+    b) branch=${OPTARG};;
+    s) since=${OPTARG};;
+    c) shallow_clone=false;;
+    o) git_options=${OPTARG};;
+    *) echo "Usage: $0 [-b branch_name] [-s since_days] [-o git_options] [-c]" >&2
+       exit 1;;
+  esac
+done
+
+# Check that both -s and -c cannot be specified at the same time
+if [ "$shallow_clone" = false ] && [ -n "$since" ]; then
+  echo "Cannot specify both -s and -c options at the same time" >&2
+  exit 1
+fi
+
+function prepare() {
+  echo -n "" > "$userListFile"
+  echo -n "" > "$userUniqueListFile"
+  echo "repository,users" > "$usersByRepoFile"
+}
+
+function print_information() {
+  echo "Counting commits from the last $since days"
+
+  if [ -z "$branch" ]; then
+    echo "No branch name specified, counting commits on default branch"
+  else
+    echo "Branch name is $branch"
+  fi
+}
+
+function process_repositories() {
+  numOfRepos=$(sed "/^[[:blank:]]*#/d;s/^[[:blank:]]*//;s/#.*//" "$repositories" | wc -l | tr -d '[:space:]')
+  current=1
+  while IFS= read -r repo
+  do
+    echo -e -n "${yellow}($current/$numOfRepos) Processing ${repo}...${nc}"
+    process_repository "$repo"
+    ((current++))
+  done < <(sed "/^[[:blank:]]*#/d;s/^[[:blank:]]*//;s/#.*//" "$repositories")
+
+  # remove duplicates in list of users
+  sort -u "$userListFile" > "$userUniqueListFile"
+}
+
+function process_repository() {
+  repository_name="${1//\//-}"
+  
+  # Check if repository is valid
+  if [ -z "$repository_name" ]; then
+    echo "Repository name is empty" >&2
+    exit 1
```

Comment:
Is exiting the entire script necessary here? The result seems to be that you get:
* The CSV file for all processed repos until the failure
* The non-unique users for every repo processed so far
* An empty unique users file

These files give the impression that the script has completed, even though it has not. It seems that maybe just returning early from this function would just skip the empty line?

Reply:
Good point.

Reply:
I removed the exit. let me know what you think.

Reply:
I think the exit needs be replaced with a `return 0` rather than omitted entirely for this to work. Otherwise, the, the git commands at the end of the function run from the `.tmp/repos` directory. This could have a few undesired outcomes:
- If this is running within a git repository, as I'm doing in build config samples here, the git commands run on that repository. So it counts committers to that outer repo and also runs a `git reset HEAD .` on that repo.
- If it is running outside of a git repo, then I would expect the git commands to error.